src/imp/mbedtls.rs

```diff
@@ -0,0 +1,487 @@
+        Ok(der)
+    }
+}
+
+fn cert_to_vec(certs_in: &[::Certificate]) -> Vec<MbedtlsCert> {
+    let mut certs : Vec<MbedtlsCert> = Vec::with_capacity(certs_in.len());
+    for cert in certs_in {
+        certs.push((cert.0).0.clone())
+    }
+    certs
+}
+
+#[derive(Debug)]
+pub struct TlsStream<S> {
+    ca_certs: *mut Vec<MbedtlsCert>,
+    ca_cert_list: *mut CertList<'static>,
+    cred_pk: *mut Pk,
+    cred_certs: *mut Vec<MbedtlsCert>,
+    cred_cert_list: *mut CertList<'static>,
+    entropy: *mut OsEntropy<'static>,
+    rng: *mut CtrDrbg<'static>,
+    config: *mut Config<'static>,
+    ctx: *mut Context<'static>,
+    session: *mut Session<'static>,
+    socket: *mut S,
+}
+
+/// ???
+unsafe impl<S> Sync for TlsStream<S> {}
+unsafe impl<S> Send for TlsStream<S> {}
+
+impl<S> Drop for TlsStream<S> {
+    fn drop(&mut self) {
+        unsafe {
+            if self.session != ::std::ptr::null_mut() {
+                Box::from_raw(self.session);
+            }
+            Box::from_raw(self.socket);
+
+            Box::from_raw(self.ctx);
+            Box::from_raw(self.config);
+            Box::from_raw(self.rng);
+            Box::from_raw(self.entropy);
+
+            if self.cred_cert_list != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_cert_list);
+            }
+            if self.cred_certs != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_certs);
+            }
+            if self.cred_pk != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_pk);
+            }
+
+            if self.ca_cert_list != ::std::ptr::null_mut() {
+                Box::from_raw(self.ca_cert_list);
+            }
+
+            if self.ca_certs != ::std::ptr::null_mut() {
+                Box::from_raw(self.ca_certs);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MidHandshakeTlsStream<S> {
+    stream: TlsStream<S>,
+    error: Error,
+}
+
+pub enum HandshakeError<S> {
+    Failure(Error),
+    WouldBlock(MidHandshakeTlsStream<S>),
+}
+
+impl<S> MidHandshakeTlsStream<S> {
+    pub fn get_ref(&self) -> &S {
+        self.stream.get_ref()
+    }
+
+    pub fn get_mut(&mut self) -> &mut S {
+        self.stream.get_mut()
+    }
+}
+
+impl<S> MidHandshakeTlsStream<S>
+where
+    S: io::Read + io::Write,
+{
+    pub fn handshake(self) -> Result<TlsStream<S>, HandshakeError<S>> {
+        panic!("what to do?");
+        Ok(self.stream) // ???
+    }
+}
+
+#[derive(Clone)]
+pub struct TlsConnector {
+    min_protocol: Option<Protocol>,
+    max_protocol: Option<Protocol>,
+    root_certificates: Vec<::Certificate>,
+    accept_invalid_certs: bool,
+    accept_invalid_hostnames: bool,
+    use_sni: bool,
+}
+
+impl TlsConnector {
+    pub fn new(builder: &TlsConnectorBuilder) -> Result<TlsConnector, Error> {
+        if builder.identity.is_some() {
+            return Err(Error::Custom("Client authentication not supported".to_owned()));
```

Comment:
Why not? Is this something that is lacking from native_tls?

Reply:
No I just need to implement this